Proposal 4.0 (HTTP/2+3, OS Trust Store, Custom DNS, OCSP, ...)

Ousret opened the pull request:
This PR showcases how HTTPie could evolve outside of Requests.

Niquests is supposed to be a (mostly) compatible fork of Requests.

Try this preview:
```shell
$ pip install "git+https://github.com/Ousret/HTTPie.git@feature-tryout-niquests" -U
```

**Here are the biggest pros of this:**
- OS truststore by default, no more certifi!
- Object-oriented headers. _Could bring additional features!_
- Fully type-annotated! 
- HTTP/3 over QUIC.
- HTTP/2 by default.
- Exit Python http.client! in favor of h11.
- Timeout by default.
- Inspect peer certificate, HTTP version, TLS version, cipher, and so on via hook/callback before a request is sent.
- All the features you expose are available in all three protocols.
- Python 3.7+, no sacrifice needed.
- Encrypted DNS support. w/ DNS-over-HTTPS, DNS-over-QUIC, DNS-over-TLS and plain DNS-over-UDP.

**Obviously, cons:**
- Stricter on emitted requests per RFCs
- Young project but based on solid bases, knowledge, and experiences.
- Need to publish new packages to distro. _Easy but time-consuming_
- Exit pyopenssl, not supported. _more of a pro to me_
- Require major bump? Could be. Should be.

Complete list of changes in the fork: https://github.com/jawah/niquests/blob/main/HISTORY.md

![Capture d’écran du 2023-11-27 19-52-12](https://github.com/httpie/cli/assets/9326700/e3d68ac9-bddc-4aaf-bf22-e0cf7a3fe5b7)

- Close #576 
- Close #480 
- Close #692 
- Close #523 
- Close #549 
- Close #1023
- Close #826
- Close #800 
- Close #1495
- Close #1202
- Close #99
- Close #1422 
- Close #94 
- Close #1480 
- Close #1522
- Close #1413
- Close #1446
- Close #414 
- Close #1551
- Close #1456
- Close #1338
- Close #1401
- Close #1527
- Close #602
- Close #962
- Close #1554 
- Close #423 
- Close #1400
- Close #1458
- Close #722
- Close #579
- Close #1580
- Close #1581
- Close #752
- Close #1599
- Close #1530

---

## [4.0.0](https://github.com/httpie/cli/compare/3.2.3...master) (unreleased)

- Switched from the [`requests`](https://github.com/psf/requests) library to the compatible [`niquests`](https://github.com/jawah/niquests). ([#1531](https://github.com/httpie/cli/pull/1531))
- Added support for HTTP/2, and HTTP/3 protocols. ([#523](https://github.com/httpie/cli/issues/523), [#692](https://github.com/httpie/cli/issues/692), [#1531](https://github.com/httpie/cli/pull/1531))
- Added support for early (informational) responses. ([#752](https://github.com/httpie/cli/issues/752)) ([#1531](https://github.com/httpie/cli/pull/1531))
- Added support for Happy Eyeballs algorithm via `--heb` flag (disabled by default). [#1599](https://github.com/httpie/cli/issues/1599) [#1531](https://github.com/httpie/cli/pull/1531)
- Added support for IPv4/IPv6 enforcement with `-6` and `-4`. ([#94](https://github.com/httpie/cli/issues/94), [#1531](https://github.com/httpie/cli/pull/1531))
- Added support for alternative DNS resolvers via `--resolver`. DNS over HTTPS, DNS over TLS, DNS over QUIC, and DNS over UDP are accepted. ([#99](https://github.com/httpie/cli/issues/99), [#1531](https://github.com/httpie/cli/pull/1531))
- Added support for binding to a specific network adapter with `--interface`. ([#1422](https://github.com/httpie/cli/issues/1422), [#1531](https://github.com/httpie/cli/pull/1531))
- Added support for specifying the local port with `--local-port`. ([#1456](https://github.com/httpie/cli/issues/1456), [#1531](https://github.com/httpie/cli/pull/1531))
- Added request metadata for the TLS certificate, negotiated version with cipher, the revocation status and the remote peer IP address. ([#1495](https://github.com/httpie/cli/issues/1495), [#1023](https://github.com/httpie/cli/issues/1023), [#826](https://github.com/httpie/cli/issues/826), [#1531](https://github.com/httpie/cli/pull/1531))
- Added support to load the operating system trust store for the peer certificate validation. ([#480](https://github.com/httpie/cli/issues/480), [#1531](https://github.com/httpie/cli/pull/1531))
- Added support for using the system trust store to retrieve root CAs for verifying TLS certificates. ([#1531](https://github.com/httpie/cli/pull/1531))
- Added detailed timings in response metadata with DNS resolution, established, TLS handshake, and request sending delays. ([#1023](https://github.com/httpie/cli/issues/1023), [#1531](https://github.com/httpie/cli/pull/1531))
- Added automated resolution of hosts ending with `.localhost` to the default loopback address. ([#1458](https://github.com/httpie/cli/issues/1458), [#1527](https://github.com/httpie/cli/issues/1527))
- Fixed the case when multiple headers where concatenated in the response output. ([#1413](https://github.com/httpie/cli/issues/1413), [#1531](https://github.com/httpie/cli/pull/1531))
- Fixed an edge case where HTTPie could be lead to believe data was passed in stdin, thus sending a POST by default. ([#1551](https://github.com/httpie/cli/issues/1551), [#1531](https://github.com/httpie/cli/pull/1531))
  This fix has the particularity to consider 0 byte long stdin buffer as absent stdin. Empty stdin buffer will be ignored.
- Improved performance while downloading by setting chunk size to `-1` to retrieve packets as they arrive. ([#1531](https://github.com/httpie/cli/pull/1531))
- Fixed multipart form data having filename not rfc2231 compliant when name contain non-ascii characters. ([#1401](https://github.com/httpie/cli/issues/1401))
- Fixed issue where the configuration directory was not created at runtime that made the update fetcher run everytime. ([#1527](https://github.com/httpie/cli/issues/1527))
- Fixed cookie persistence in HTTPie session when targeting localhost. They were dropped due to the standard library. ([#1527](https://github.com/httpie/cli/issues/1527))
- Fixed downloader when trying to fetch compressed content. The process will no longer exit with the "Incomplete download" error. ([#1554](https://github.com/httpie/cli/issues/1554), [#423](https://github.com/httpie/cli/issues/423), [#1527](https://github.com/httpie/cli/issues/1527))
- Fixed downloader yielding an incorrect speed when the remote is using `Content-Encoding` aka. compressed body. ([#1554](https://github.com/httpie/cli/issues/1554), [#423](https://github.com/httpie/cli/issues/423), [#1527](https://github.com/httpie/cli/issues/1527))
- Removed support for preserving the original casing of HTTP headers. This comes as a constraint of newer protocols, namely HTTP/2+ that normalize header keys by default. From the HTTPie user perspective, they are "prettified" in the output by default. e.g. `x-hello-world` is displayed as `X-Hello-World`.
- Removed support for `pyopenssl`. ([#1531](https://github.com/httpie/cli/pull/1531))
- Removed support for dead SSL protocols < TLS 1.0 (e.g. sslv3) as per pyopenssl removal. ([#1531](https://github.com/httpie/cli/pull/1531))
- Removed dependency on `requests_toolbelt` in favor of directly including `MultipartEncoder` into HTTPie due to its direct dependency to requests. ([#1531](https://github.com/httpie/cli/pull/1531))
- Removed dependency on `multidict` in favor of an internal one due to often missing pre-built wheels. ([#1522](https://github.com/httpie/cli/issues/1522), [#1531](https://github.com/httpie/cli/pull/1531))

Existing plugins are expected to work without any changes. The only caveat would be that certain plugin explicitly require `requests`.
Future contributions may be made in order to relax the constraints where applicable.

Ousret commented:
update:

![image](https://github.com/httpie/cli/assets/9326700/55808eb0-db09-4b91-a3f8-cfc6fd92f914)

- Niquests support universal revocation verification in addition to the native OS root CAs.
- Reviewed output format for debug info to match HTTPie's format.
- Down to 9 failures, which seems trivial enough to fix.

Ousret commented on tests/test_json.py:
```diff
@@ -334,23 +334,24 @@ def test_complex_json_arguments_with_non_json(httpbin, request_type, value):
-                },
-            },
-        ),
+        # todo: this expression leads to an invalid header name! this is a bug.
```
FYI this test case had to be disabled. It produces an invalid header name.
requests did not say anything alarming due to the too-permissive underlying urllib3 x http.client.

Ousret commented:
The remaining 9 cases run fine if standalone. something tampers with the exception management. thus making assertions fail consistently. need investigation.

Ousret commented:
Update:

Now everything has passed. I identified some parts of the code that could be removed.

- `https GET pie.dev/get`
![image](https://github.com/httpie/cli/assets/9326700/b2c3d2f4-b164-4d27-b6e6-4132ee0ce97e)
- `https -v GET pie.dev/get`
![image](https://github.com/httpie/cli/assets/9326700/8dbb273e-3d4c-40d8-89aa-238a94165218)
- `https -v --debug GET pie.dev/get`
![image](https://github.com/httpie/cli/assets/9326700/27f8cd89-efd9-4c96-8a8d-4a8d33f459c1)
- `https --ssl=tls1.2 -v --debug GET pie.dev/get`
![image](https://github.com/httpie/cli/assets/9326700/fcf03313-9746-4440-9344-1e35e5266e1a)

- `http -v --debug GET pie.dev/get`
![image](https://github.com/httpie/cli/assets/9326700/ef6a7f1f-5313-48bc-b038-356d2ad3f676)

Ousret commented:
Tests / test (macos-latest, 3.11) (pull_request) Failing after 2m
[Details](https://github.com/httpie/cli/actions/runs/6510806518/job/17685266648?pr=1531)
@github-actions
Tests / test (macos-latest, 3.12) (pull_request) Failing after 4m

it is about CI flakyness, a restart will fix that.

Ousret commented:
So far, I identified this:

- Close #576 
- Close #480 
- Close #692 
- Close #523

Ousret commented:
We can add the following arguments and document them:

- `--disable-http2`
- `--disable-http3`
- `--http3`, e.g. force trying it.

Ousret commented:
I've added

- `--disable-http2`
- `--disable-http3`
- `--http3`, e.g. force trying it.

Ideally, the connection information (TLS, Cipher, Peer/Issuer Certificate, and OCSP) should be in the request metadata and with a lexer instruction.

dwt commented:
Damn this looks so tantalizing. I'd love for this to get packaged as a tryout / beta. Something installable via e.g. `pip install httpie[niquest]` perhaps?

Ousret commented:
Hello there,

---

@dwt 

> Damn this looks so tantalizing.

Couldn't agree more, wait until you see what just landed in this PR.

> I'd love for this to get packaged as a tryout / beta.

That's the main idea I had in mind.
If you or anyone you know is willing to give this a try, then feel free to share and ask for feedbacks.

---

@jkbrzt 

I, again, am still hopeful you could spare a bit of time to steer this proposal.
With the most recent push, there's some interesting news.

Starting now, HTTPie PR/1531 can:

- Use custom DNS resolver, with but not limited to: #99
  - DNS over HTTPS
  - DNS over QUIC
  - DNS over TLS
  - DNS over UDP

```bash
$ https --resolver "doh+cloudflare://" pie.dev/get
```

- Bind to a specific network interface. #1422

```bash
$ https --interface 172.17.0.1 pie.dev/get
```

- Enforce IPv4, or IPv6. #94

```bash
$ https -4 pie.dev/get
```

- Detailed timings in response metadata with DNS resolution, established, TLS handshake, and request sending delays. #1023

![image](https://github.com/httpie/cli/assets/9326700/8fb796d7-d52b-4a25-aa78-be321a25f356)

This is by far, according to me, the most exiting upgrade HTTPie can land and justify the major bump.
I took the liberty to draft a changelog for a "potential" 4.0.0 beta 1.

Regards,

dwt commented:
Well, that would be the most interesting update to httpie in some time according to me too. :)

dwt commented:
@Ousret I'm trying out the http3 support and wasn't able to get it working. Could you perhaps provide some guidance on how to best debug this?

```shell
❯ https --http3 -vvv pie.dev/get
Connected to: 188.114.96.4 port 443
Connection secured using: TLSv1.3 with AES-256-GCM-SHA384
Server certificate: commonName="pie.dev"; DNS="*.pie.dev"; DNS="pie.dev"
Certificate validity: "Nov 11 01:14:24 2023 GMT" to "Feb  9 01:14:23 2024 GMT"
Issuer: countryName="US"; organizationName="Let's Encrypt"; commonName="E1"
Revocation status: Good

GET /get HTTP/2
Accept: */*
Accept-Encoding: gzip, deflate
Connection: keep-alive
Host: pie.dev
User-Agent: HTTPie/4.0.0.b1
```
<details><summary>Details</summary>
<p>
```shell
❯ https --http3 -vvv pie.dev/get
Connected to: 188.114.96.4 port 443
Connection secured using: TLSv1.3 with AES-256-GCM-SHA384
Server certificate: commonName="pie.dev"; DNS="*.pie.dev"; DNS="pie.dev"
Certificate validity: "Nov 11 01:14:24 2023 GMT" to "Feb  9 01:14:23 2024 GMT"
Issuer: countryName="US"; organizationName="Let's Encrypt"; commonName="E1"
Revocation status: Good

GET /get HTTP/2
Accept: */*
Accept-Encoding: gzip, deflate
Connection: keep-alive
Host: pie.dev
User-Agent: HTTPie/4.0.0.b1



HTTP/2 200 OK
Access-Control-Allow-Credentials: true
Access-Control-Allow-Origin: *
Alt-Svc: h3=":443"; ma=86400
Cf-Cache-Status: DYNAMIC
Cf-Ray: 83fa5064bee41d9a-FRA
Content-Encoding: gzip
Content-Type: application/json
Date: Wed, 03 Jan 2024 09:46:20 GMT
Nel: {"success_fraction":0,"report_to":"cf-nel","max_age":604800}
Report-To: {"endpoints":[{"url":"https:\/\/a.nel.cloudflare.com\/report\/v3?s=Ca0pVbLbx9QmhpCG%2Fq03dTYDMeHurFMramSLFHceVhaq6dMGWBep340o4G9B3t%2FlHxfF5Agpp7UHgkxGa5AMFVyvov%2BhOGIsDikuFWJMVFWp4lN9Xnd%2FILz9"}],"group":"cf-nel","max_age":604800}
Server: cloudflare

{
    "args": {},
    "headers": {
        "Accept": "*/*",
        "Accept-Encoding": "gzip",
        "Cdn-Loop": "cloudflare",
        "Cf-Connecting-Ip": "91.65.247.116",
        "Cf-Ipcountry": "DE",
        "Cf-Ray": "83fa5064bee41d9a-FRA",
        "Cf-Visitor": "{\"scheme\":\"https\"}",
        "Connection": "Keep-Alive",
        "Host": "pie.dev",
        "User-Agent": "HTTPie/4.0.0.b1"
    },
    "origin": "91.65.247.116",
    "url": "https://pie.dev/get"
}


Elapsed DNS: 0.002622s
Elapsed established connection: 0.026514s
Elapsed TLS handshake: 0.079068s
Elapsed emitting request: 0.000366s
Elapsed time: 0.254063667s⏎                                                                                                        
~/C/P/httpie  🐍 tempenv-13fd235984bf4 🌱 feature-tryout-niquests  
❯ macosver
10.16

~/C/P/httpie  🐍 tempenv-13fd235984bf4 🌱 feature-tryout-niquests  
❯ python --version
Python 3.12.1

~/C/P/httpie  🐍 tempenv-13fd235984bf4 🌱 feature-tryout-niquests  
❯ pip list
Package            Version    Editable project location
------------------ ---------- -------------------------------
certifi            2023.11.17
cffi               1.16.0
charset-normalizer 3.3.2
cryptography       41.0.7
defusedxml         0.7.1
h11                0.14.0
h2                 4.1.0
hpack              4.0.0
httpie             4.0.0b1    /Users/dwt/Code/Projekte/httpie
hyperframe         6.0.1
idna               3.6
kiss-headers       2.4.3
markdown-it-py     3.0.0
mdurl              0.1.2
multidict          6.0.4
niquests           3.4.0
pip                23.3.2
pycparser          2.21
Pygments           2.17.2
PySocks            1.7.1
python-socks       2.4.4
qh3                0.14.0
requests           2.31.0
requests-toolbelt  1.0.0
rich               13.7.0
setuptools         69.0.3
urllib3            2.1.0
urllib3-future     2.4.902
wassima            1.0.3
```
</p>
</details>

Ousret commented:
OK, I can try.
First, running the exact same command locally, work fine.

You don't have to specify `--http3` as pie.dev is covered by Cloudflare preemptive QUIC declaration in DNS records.

According to your dependencies listening, you have qh3 installed, so it should work and everything is up to date.

Something is preventing it on your environment. Let's try to understand. Are you using conda? or poetry? or..?

Try to run the following, standalone:

```python
from niquests import Session

with Session() as s:
    print(s.get("https://pie.dev/get"))
```

and then:

```python
from urllib3.contrib.hface import HTTPProtocolFactory, HTTP3Protocol

print(HTTPProtocolFactory.new(HTTP3Protocol))
```

Ousret commented:
Thanks to @dwt testing, we discovered an edge case where http3 would be silently disabled, it depended on the ssl ctx created with the system default ciphers list, which vary.
With this, I discovered a unusual bug that is present in the 3.x major, see https://github.com/httpie/cli/issues/1551 for more details.
It is fixed now (both cases).

Ousret commented:
Now, this fixes 20 issues. I like round numbers. Just added support for `--local-port` args. Accept single port or range.

Ousret commented:
gentle ping @jkbrzt

dwt commented:
@jkbrzt hey, this code seems like a phenomenal update to the capabilities of httpie. Could you please take a look or recommend someone else you trust to take one?

dwt commented:
@Ousret It seems to be hard to reach @jkbrzt - any Idea how to get him to notice? I didn't find any social addresses outside of twitter, and it seems they want me to pay now to send private messages there.

Ousret commented:
I understand the frustration.

There's almost no chance that @jkbrzt didn't saw this (amongst other PR and issue piling up of course), so there's really no reason for us to ping him any more. This PR is 4 months old already, and we didn't get the slightest sign of whether it is a desirable PR or not.

According to stats at my disposal we have a significant amount of "git clone" that must originate from people that can't setup HTTPie with Python 3.12+ around 10 a week, and I expect that number to quickly raise, this is undesirable, maintenance wise.

I tend to wait 5 to 6 months before giving up on a PR, usually. If, HTTPie would be stuck this way, someone would have to assume a fork, and, well, it won't be me as I have no bandwidth to allocate for long term maintenance.

dwt commented:
I have to admit that I'm already very impressed that you invested all of. this work _without_ having any word from the project owner whether he likes this or not. I sure hope that we can get him to reply at some point, because this looks like a big present he got for free for this project.

dwt commented:
@jkbrzt How can we reach you? It would be really nice (but after this long a time also neccessary for you to chime into this pull request and give some information what you would require to get this merged.

jkbrzt commented:
@Ousret Impressive stuff, and sorry for the lack of feedback. The initial effort looked like a speculative experiment, but it’s evolved quite a bit since then! I have some questions. Would you be up for connecting over Zoom/Meet to discuss the code, niquests, and the potential switch from requests? If it sounds good, please shoot me a message via jakub at httpie.io.

Ousret commented:
Made some general (minor) code cleanup and fixed almost all warnings.

Passed from (master) `===== 1019 passed, 5 skipped, 4 xfailed, 135 warnings in 71.36s (0:01:11) ======`

to (this PR) `====== 1037 passed, 5 skipped, 1 xfailed, 5 warnings in 77.09s (0:01:17) =======`

The 5 warnings remaining are going to be annoying to track down and fix for now.

jkbrzt commented:
@Ousret thank you for the call yesterday. 

I’ve started diving into the code and doing some manual QA. 

Here are a few cosmetic issues I’ve noticed so far:

```bash
# Test command
$ https -vv pie.dev/get
```

- [ ] Request-Line and Status-Line are no longer syntax-highlighted in the output. The HTTP Pygments lexer probably needs to be updated for HTTP > 1.

- [ ] We need an extra blank line after the `--vv` output. Currently the output is `Elapsed time: 0.118802041s$` where `$` is my shell prompt.

- [ ] We need consistent and user-friendly rounding for various durations. The download progress probably has something usable we could apply everywhere. (The current output sometimes uses scientific notation: `Elapsed established connection: 4.8e-05s`)

<img width="1857" alt="httpie3vs4" src="https://github.com/httpie/cli/assets/326885/a9941d77-fbd3-49b2-abd1-2b733cf604c5">

Ousret commented:
OK! The cosmetic issues are now fixed. I took the liberty to add a single assert that verify we've not missed the extraneous RC for the metadata.

For now, all clear.

Ousret commented:
Did found another cosmetic issue (when follow redirects) the meta was glued to the next request line.
Now fixed. And made the metadata look even nicer by replacing "0.0s" with "0s". (Found it in same scenario where redirect did not require DNS resolve and such..)
Finally, I could resolve 4 warnings, remain 1! `======= 1039 passed, 5 skipped, 1 xfailed, 1 warning in 77.85s (0:01:17) =======`

Ousret commented:
So, I did some additional testing on my end and found nothing.
Additionally, to mark this milestone, the very last warning is now fixed!

This branch now have 0 warning. `============ 1039 passed, 5 skipped, 1 xfailed in 81.02s (0:01:21) =============`

For a side note, [Niquests](https://github.com/jawah/niquests/releases/tag/v3.5.5) just landed [Happy Eyeballs](https://niquests.readthedocs.io/en/latest/user/quickstart.html#happy-eyeballs) feature in the last version, if this is of any interest to you, let me know.

Ousret commented:
I applied a minor patch to improve how the QuicCapabilityCache blend into HTTPie internals (the path is now generated within env::config and propagated appropriately). Tests added to ensure no regression happen in between.

dwt commented:
@jkbrzt @Ousret is there anything I can do to help test this more? I really want to see this land. :)

Ousret commented:
@dwt You can help by proof reading the modified documentation and look if I missed something regarding this PR in it: https://github.com/Ousret/httpie/blob/feature-tryout-niquests/docs/README.md

Other than that, I think the testing part is pretty much covered.

Finally I have added a test case for the following scenario: "A peer was H3 compatible but not anymore and expect a error msg that explain why you just got timeout and how to remedy".

Ousret commented:
Issue https://github.com/httpie/cli/issues/1401 was trivial to fix, so, it's fixed and tested.

dwt commented:
> @dwt You can help by proof reading the modified documentation and look if I missed something regarding this PR in it: https://github.com/Ousret/httpie/blob/feature-tryout-niquests/docs/README.md

I've had a look at the docs and so far everything seems fine. However I noticed the absence of a `--http2` flag to force http2. Could you touch on the rationale behind that omission? I am currently debugging a server who has trouble spitting out correct http2, so this would be a useful feature to me?

dwt commented:
I may be missing something here, but does that imply that the client has no way to force http1 2 or 3 according to its wishes? When setting up server support for these protocols, I would like to have the ability to force all of these protocols for debugging purposes?

Ousret commented:
Just updated the docs to include previous comment about protocol combo and the availability of HTTP/3 or not.
Since last time Niquests made HTTP/3 support far more enjoyable with less dependencies and a faster execution time.

Finally, we had to downgrade macos-14 (macos-latest) to macos-13 because it's relatively new and some dependencies like brotlicffi aren't ready for it.

mshahat commented:
Hello @Ousret , nice work here ... is there an ETA on when to expect this release to become available? Thank you

Ousret commented:
No precise ETA. We are waiting upon the owner approval.

markstos commented:
Looks like it could be a big step forward if @jkbrzt approves it. Last week I wanted to test resolving over IPv4 vs IPv6. I found that the released version could not help with this task, but it worked perfectly in the fork, with the same `-4` and `-6` flags that `dig` uses.

Ousret commented:
> Excited to see HTTP/2 and HTTP/3 soon landing in HTTPie 🎉

Likewise!

> I was wondering if we could have an equivalent to cURL's http-version flags

I first need to add a feature into Niquests, which isn't top priority right now.
But it is planed, will be in a future minor. Provided everything goes smooth with this.

regards,

janbrasna commented on docs/README.md:
```diff
@@ -1530,13 +1537,13 @@ $ http --cert=client.pem --cert-key=client.key --cert-key-pass=my_password https
 ### SSL version
 
 Use the `--ssl=<PROTOCOL>` option to specify the desired protocol version to use.
-This will default to SSL v2.3 which will negotiate the highest protocol that both the server and your installation of OpenSSL support.
-The available protocols are `ssl2.3`, `ssl3`, `tls1`, `tls1.1`, `tls1.2`, `tls1.3`.
+This will default to TLS v1.0 which will negotiate the highest protocol that both the server and your installation of OpenSSL support.
+The available protocols are `tls1`, `tls1.1`, `tls1.2`, `tls1.3`.
```
Love the idea, and without digging any deeper into the changeset to verify my thoughts in implementation, I'm just raising it here for general discussion:

"This is will default to TLS v1.0 which will negotiate the highest protocol that both […] support."

1. This sounds a bit confusing. It's either "This is will default to TLS v1.0" (which it should not), or "This is will default to a highest TLS version that both […]".
 - Because the values don't mean "use as the lowest", but "use the chosen only" (i.e. not a tls12+ or a range, but tls12 _only_)
 - I'd actually propose a magic value like "`tls`" (roughly meaning an alias to `PROTOCOL_TLS_CLIENT`) that would use the highest protocol available. Not sure if that's the default thou, not sure if the client can have tls13 capability, but still use tls12 as a default? If default == max always, then there's prolly no need for such value, only way to use max protocol then is to omit the param though. Which would mean no "max" magic value would be listed in help/man among the "available protocols", which in turn might seem a bit confusing then. So an explicit value for max protocol available, even when it really doesn't do anything if that's the default, can still have its value for clarity.
  - Do you plan to silently accept the old "`ssl2.3`" value mapped to default, whatever that ends up to be, not to raise any errors?

3. Are you sure the "`tls1.3`" value works? It does not in master, and never did (since the openssl apis are different for TLSv13, and the change in flags structure [see 4. ↓] made it impossible to constrain without all `*_OP_NO_*` used) — it only selects TLSv13 if that would have been the default anyways…

4. Adding ranges, i.e. min=tls11 max=tls12, or just one constraint alone (to match the current python sslcontext, instead of selecting one version, or disabling individual one by one) would also be lovely at some point.

Ousret replied:
Indeed, the "This is will default to ..." is kind of confusing. I've addressed that by rephrasing it.

1) It is rephrased accordingly with "If not specified, it tries to negotiate the highest protocol that both the server and your installation of OpenSSL support."
2) "tls1.3" case has been addressed, we are now assuming that "tls1.3" (associated to PROTOCOL_TLS_CLIENT) means strictly do tls 1.3, and appropriate measures have been made.
3) The range of accepted tls version should be dealt with separately, I don't want to further increase this PR size, thus delaying it further on. I agree it can be useful, but it's not top priority as we speak.

 regards,

Ousret replied:
To demonstrate `--ssl=tls1.3` is working, run `https --ssl=tls1.3 https://tls-v1-2.badssl.com:1012/`

janbrasna commented on extras/man/http.1:
```diff
@@ -570,9 +570,10 @@ installation (only the supported ones are shown here).
 
 
 A string in the OpenSSL cipher list format.
+tls1.3 ciphers are always present regardless of your cipher list.
```
This may need a bit more context or outsider-friendly rewording. WDYT?

``` suggestion
Applies only for TLS v1.2 and lower. TLS v1.3 connections always use the appropriate cipher suites available for that protocol version, regardless of this setting.
```

Ousret replied:
see https://docs.python.org/3/library/ssl.html#ssl.SSLContext.set_ciphers

we can't do anything about it. unless we're using http3 over quic.

janbrasna replied:
I'm not challenging anything from that. Most APIs are only handling tls12 ciphers, and don't bother with lower-level ctxs to set tls13 suites, esp. in OpenSSL world. — I'm only raising the need to explain that to consumers not familiar with the distinction in the new protocol RFC.

Ousret replied:
Of course, there's certainly a lot of room to improve the wording for a wider range of users. The primary goal of this PR is to move forward on the feature level and to detach ourselves from Requests.
Follow up PRs are of course, welcomed.

janbrasna commented on extras/man/https.1:
```diff
@@ -570,9 +570,10 @@ installation (only the supported ones are shown here).
 
 
 A string in the OpenSSL cipher list format.
+tls1.3 ciphers are always present regardless of your cipher list.
```
_(same as above)_

janbrasna commented on httpie/cli/definition.py:
```diff
@@ -906,12 +906,12 @@ def format_auth_help(auth_plugins_mapping, *, isolation_mode: bool = False):
 CIPHERS_CURRENT_DEFAULTS = (
     """
-    See `http --help` for the default ciphers list on you system.
+    See `http --help` for the default ciphers list.
```
_(same as above)_

janbrasna commented on httpie/cli/definition.py:
```diff
@@ -906,12 +906,12 @@ def format_auth_help(auth_plugins_mapping, *, isolation_mode: bool = False):
     if IS_MAN_PAGE else
     f"""
-    By default, the following ciphers are used on your system:
+    By default, the following ciphers are used:
```
_(same as above)_

janbrasna commented on httpie/cli/definition.py:
```diff
@@ -923,6 +923,7 @@ def format_auth_help(auth_plugins_mapping, *, isolation_mode: bool = False):
     help=f"""
 
     A string in the OpenSSL cipher list format.
+    tls1.3 ciphers are always present regardless of your cipher list.
```
_(same as above)_

janbrasna commented on httpie/ssl_.py:
```diff
@@ -104,6 +104,18 @@ def __init__(
+                # Please do not raise a "security" concern for that line.
+                # If the interpreter reach that line, it means that the user willingly set
+                # an unsafe TLS protocol.
+                ciphers = "DEFAULT:@SECLEVEL=0"
```
Q: just a quick check — the case of TLSv1(.1) with explicit --ciphers is not covered here; wouldn't you need the seclevel added for these as well, or you'd expect the user to add the at-rule themselves in that case?

Ousret replied:
If the user put a cipher list himself, we should not interfere. HTTPie will interfere (user knows what he's doing)
adding SECLEVEL appended to his suites is going to bring confusion.

Ousret replied:
short answer: yes we should let the user decide, or/and figure it out himself.

janbrasna replied:
Q: I've learned the ciphers are not DEFAULT but own constant set; yet here the seclevel string appends to a DEFAULT. is that what is expected?

janbrasna reviewed (commented):
A couple of questions regarding how to surface some of the functionality and constraints to the users:

Ousret commented on extras/man/http.1:
```diff
@@ -570,9 +570,10 @@ installation (only the supported ones are shown here).
 
 
-See `http \fB\,--help\/\fR` for the default ciphers list on you system.
+See `http \fB\,--help\/\fR` for the default ciphers list.
```
because it no longer use the default cipher list provided by the system.
In fact, we restore the original behavior that was removed by urllib3 (v1 to v2) where the cipher list was a constant.
we started using a consistent cipher list, using the mozilla recommended cipher suites.

janbrasna replied:
ah cool!

janbrasna replied:
in that case, I'm wondering if there's a good place generally, somewhere httpie top-level, to explain that the client cipher list available is not the system/openssl/context default, but httpie now has its own (opinionated) list that may or may not be the subset of what other cli/connections on such system are able to negotiate, which may lead to unexpected results now.

FYI the [wiki.mozilla.org/Security/Server_Side_TLS](https://wiki.mozilla.org/Security/Server_Side_TLS#Intermediate_compatibility_.28recommended.29) cipher lists are meant to be server-side hardening, not client-side preference, so it may be unnecessarily restraining for a general http client w/o any prior heads up.

disclaimer: I use httpie to debug api connections where cipher + protocol (mis)match / handshake failure is common so having a different client cipher list than what the system has set (not just by default, but even intentionally by configuration in some distros)  is troublesome if it's not brought to my attention in advance.
disclaimer2: I'm involved with @mozilla TLS/SSL infosec so I might be a bit biased, but at the same time I'm well aware there are cipher suites not on the lists that are otherwise completely fine to be used if you know the context, just for the matter of simplification and automating config outputs only the "popular" enough (read: conservative and necessary) are being kept in the list, which is somewhat a dumbed-down subset compatible enough for those not able to turn the knobs a tweak the hardening tuning themselves, that can be used as-is without much understanding the impacts. So it misses CCMs, PSKs and such and that may surprise some users expecting pre-existing implicit defaults/configs on their OS level.

Ousret replied:
> but httpie now has its own (opinionated) list that may or may not be the subset of what other cli/connections on such system are able to negotiate, which may lead to unexpected results now.

I understand your concern. There's concern on my side about your reasoning.
HTTPie did abruptly switch that behavior in a minor version due to urllib3 v2 removing "DEFAULT_CIPHERS" and there was some complaints, but not in HTTPie (aside from the breakage) as far as I have seen.

> FYI the [wiki.mozilla.org/Security/Server_Side_TLS](https://wiki.mozilla.org/Security/Server_Side_TLS#Intermediate_compatibility_.28recommended.29) cipher lists are meant to be server-side hardening

I know about that. Meaning the original purpose.

My thinking / reasoning is that we need to have a cipher list that is "strong enough" and exclude de facto weak suites, and this suite, I trust. It's safe enough and permit a broad compatibility with TLS 1.2+
We live in the beginning of the "quantum cpus" era and we must be careful moving forward to protect ourselves, we've seen things no one dared to imagine with SSLv2, TLSv1 and TLSv1.1. So we must not trust too blindly the scenario "propose a wide range of ciphers" and expect the highest common cipher will be strong enough.

> list that may or may not be the subset of what other cli/connections on such system are able to negotiate

do we have a list of popular tools and the choice around ciphers?

> so having a different client cipher list than what the system has set (not just by default, but even intentionally by configuration in some distros) is troublesome if it's not brought to my attention in advance ... just for the matter of simplification and automating config outputs only the "popular" enough ...

My humble experience say the exact opposite happen very often, meaning that user does complaint with: "help! work on my mac X but not on my brand new mac Z" or "work on Linux Debian not on Fedora" with "but, same HTTPie version!" and so on. But your personal case is different.

> So it misses CCMs, PSKs and such and that may surprise some users expecting pre-existing implicit defaults/configs on their OS level.

And, maybe I am mistaken, but I am willing to bet no one actually know the extend of their "default" ciphers list (openssl lib) and how they work. So, here, actually HTTPie will offer more consistency.
Moreover, the most surprising is that users are getting blocked by popular WAF like Cloudflare if they advertise such ciphers.

If you have a way to rephrase the "help/hint/docs" for a broader audience, we accept suggestions. of course.
or maybe having a `--system-ciphers` or similar could do the trick.

regards,

Ousret commented:
In the latest commit:

- Added a way to force http2, http1 
- So, now, we can enforce http2 by disabling http1
- http2 with prior knowledge is supported
- fixed the "real" speed download measured and reported when remote compressed the body

MattElek commented:
It's now been over a year since this pull was opened, and almost three months since the last activity. I've been using this at least six months with no issue, installed locally with pip. What is the holdup on seeing this merged??

markstos commented:
If  there is a concern that something will break for someone: yes it will, a lot has changed. But at this point the fastest way to find and resolve those edge cases may be to  go ahead and release to find out what those edge cases are. On the whole there are a lot of good changes here and the positives seem like they outweigh the negatives.

dwt commented:
I mean, if we want a staged rollout, this could be released as a beta release. If this is promoted on the homepage, I'm sure we will get much more coverage.

Ousret commented:
Hello,

Long time since I have look at this PR.
Unfortunately I did not heard back from the maintainer.

Since last time, Niquests evolved quiet a lot.

The client now support:
- Early responses
- Trailers (headers after response)
- Post-Quantum Security (Key Exchange QUIC)
- WebSocket support
- Automatic downgrade of protocol in case of known issues
- etc...

I've added in this PR:
- Support for early responses (1xx informational)
- Python 3.13 testing in CI

I agree with you all that the best way to move forward is to cut a release and listen closely to the feedback and act iteratively.

regards,

dwt commented:
@jkbrzt Aren't you the maintainer here? Could you give an update on this?

Ousret commented:
Just fixed the last sneaky warning in the tests. The suite now complete perfectly.
Took a bit of time and added Happy Eyeballs support via `--heb` flag to help users like in #1599 , through I did not enable it by default. will think of later to do so.
Also added a preventive fix to enforce all plugins use Niquests (transport & such) to ensure the highest level of compatibility between v3 & 4.


this PR fixes 35 issues[...] now.

techknowlogick commented:
@Ousret, like the others in the thread, I also wanted to thank you for your work on this PR. I have been building it with my NixOS flake https://gitea.com/techknowlogick/tklk-flake as I run the PR version daily.

markstos commented:
@Ousret At what point do you go ahead and publish your work as a fork?

Ousret commented:
Unfortunately this PR is no where near getting merged.
This has been more than a year and we do not came really closer to merging this.

I am sorry for the folks that were waiting patiently for this.
I had the chance to privately speak with the maintainer and he have a major concern that can be summarized as follow:

> "HTTPie is a UX focused tool, not supposed to be a swiss-army knife like cURL. And my testing show significant slowdown when issuing a request, passing from 400ms to 600ms. There was a ton of complaint in the past about HTTPie performance, so it is to me very important."

The performance does get a bit slower due to introduced complexity. https://github.com/Ousret/httpie-ckop/actions/runs/11548333983 And he his right, we have a solid 50% slower on a magnitude of milliseconds. I do not personally share his point of view on it through, for me it is almost imperceptible. But it does not mean I am right either. My philosophy there is that we can't get everything perfect, with no downside. Otherwise it would be a scam.

- HTTP/2, and HTTP/3 requires more computational power to execute a single query and retrieve the response https://niquests.readthedocs.io/en/latest/community/faq.html#why-http-2-and-http-3-seems-slower-than-http-1-1
- The import / module tree is bigger, thus making the interpreter slower at startup.

I understand how frustrating this is. Believe me, if I could make this faster, I would have done so.
For me, that PR remained open too long, and is dragging ~10 installs per week and with recurrent ping in it, this has to stop now for my own sanity.

To be clear, the maintainer did not refuse this contribution, more like isn't ready to move on this. And he is perfectly entitled to his choices with his project.

It could take several months maybe or never. I don't know. But I find it preferable to keep hope to zero, so that the disappointment would be reduced to the minimum. If the author come forth, he will do so in his own terms and timeline.

If someone is willing/have the courage to make this live in a separate fork, it's your choice.
I am already at maximum charge for open source projects.

Thanks for the numerous tester out there.

Regards,

markstos commented:
@Ousret Thank you for all your work on this.

As a `httpie` user, I'm also interested in "correctness over performance". I would rather be able to test HTTP/2 and HTTP/3 connections a fraction of a second slower than to not be able to test them at all.

I arrived at this pull request because I was affected by one of the amount two dozen issues that it closes. If the original maintainer doesn't merge it, I hope it's packaged as a fork. Though, my own open source plate is also full as well, and I'm not volunteering at the moment, either.

jkbrzt commented:
Sorry to see this. I expressed concerns over performance as with Niquests, the most common use case of fetching a JSON response is 50% slower. My last words were that I'll do more testing. For the record, we have been in touch over email (as well as sponsoring Ahmed) for over half a year. I wish all the best to both Ahmed and his project.
